### PR TITLE
rgw: fix bucket index list test error

### PR DESCRIPTION
**Pull request by ivancich**
Fix bug surrounding calculation of number of entries returned. Add additional level 20 logging. 

In test, make sure marker is cleared. Put end-of-list check inside the conditional with the rest of the test. End use of uninitialized variable.

Signed-off-by: J. Eric Ivancich <ivancich@redhat.com>

Fixes: https://tracker.ceph.com/issues/52315

### CODE REVIEW

File: src/cls/rgw/cls_rgw.cc [hunk 1]
```diff
@@ -384,14 +384,18 @@ static void split_key(const string& key, list<string>& vals)
...
+static std::string escape_str(const std::string& s)
 {
   int len = escape_json_attr_len(s.c_str(), s.size());
   std::string escaped(len, 0);
   escape_json_attr(s.c_str(), s.size(), escaped.data());
   return escaped;
 }
 
+static std::string quote_escape(const std::string& s) {
+  return '"' + escape_str(s) + '"';
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw.cc:396]: perhaps its being fussy, but perhaps pre-allocate or reserve escape_str?

Comment (ivancich) [reply src/cls/rgw/cls_rgw.cc]: Not sure if the change is what you had in mind.

File: src/cls/rgw/cls_rgw.cc [hunk 2]
```diff
@@ -2586,19 +2601,20 @@ static int list_plain_entries_help(cls_method_context_t hctx,
     try {
       decode(e, biter);
     } catch (ceph::buffer::error& err) {
-      CLS_LOG(0, "ERROR: %s: failed to decode buffer for plain bucket index entry \"%s\"",
-	      __func__, escape_str(iter.first).c_str());
+      CLS_LOG(0, "ERROR: %s: failed to decode buffer for plain bucket index entry %s",
+	      __func__, quote_escape(iter.first).c_str());
```
Comment (cbodley) [inline src/cls/rgw/cls_rgw.cc:2605]: what's wrong with quotes in the format strings? i prefer the existing version, especially when `quote_escape()` has runtime overhead

File: src/cls/rgw/cls_rgw.cc [hunk 3]
```diff
@@ -2635,7 +2645,7 @@ static int list_plain_entries(cls_method_context_t hctx,
 {
   int r;
   bool end_key_reached;
-  bool more;
+  bool more = false;
```
Comment (cbodley) [inline src/cls/rgw/cls_rgw.cc:2665]: can we please initialize `r` and `end_key_reached` while we're at it?

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
https://pulpito.ceph.com/cbodley-2021-09-02_14:03:39-rgw-wip-cbodley-testing-distro-basic-smithi/

seeing a lot of new failures there

**Comment by ivancich** [comment]
Teuthology results:

rgw/verify: https://pulpito.ceph.com/ivancich-2021-09-03_00:17:55-rgw-wip-fix-bi-test-error-distro-basic-smithi/
rgw/(non-verify): https://pulpito.ceph.com/ivancich-2021-09-03_00:18:24-rgw-wip-fix-bi-test-error-distro-basic-smithi/

I do not believe there are any regressions. These tests were run with --suite-branch set to the branch, so the new tests were used.

@cbodley Is it possible that the failures you saw were due to another PR on your branch or the present state of master?

**Comment by cbodley** [comment]
thanks @ivancich!

**Pull request merged**